### PR TITLE
Delay Cross Spectrum Estimator

### PR DESCRIPTION
**Pull request by jrs65**
- feat(testing): add a routine for generating test frequency spectra
- perf(random): add a multithreaded random number generator
- refactor(delay): remove redundant window_generalised implementation
- refactor(DelaySpectrumEstimatorBase): remove unnecessary setup method
- refactor(DelaySpectrumEstimatorBase): cleanup of internal logic
- feat(testing): add a task to generate random correlated sidereal streams
- feat(containers): add DelayCrossSpectrum container
- feat(delay): add routine for estimating a delay cross power spectrum
- fix: sym_pos argument to la.solve has been removed
- feat(DelaySpectrumEstimatorBase): cross spectrum estimation

### CODE REVIEW

**Comment by jrs65** [comment]
All the tests are failing due to this Cython issue, but this is the PR. I guess I'll try and fix the Cython failures.

File: draco/util/_fast_tools.pyx [hunk 1]
```diff
@@ -260,3 +271,207 @@ ctypedef fused real_or_complex:
...
+
+    cdef real_or_complex ti, tj
+
+    if out.shape[0] != nd or out.shape[1] != nd:
+        raise ValueError(f"Out array does not have the required shape ({nd}, {nd}).")
+
+    cdef real_or_complex* means = <real_or_complex*>malloc(sizeof(real_or_complex) * nd)
+    cdef real_or_complex* tarr = <real_or_complex*>malloc(sizeof(real_or_complex) * nd)
+
+    #for ii in prange(nd, nogil=True):
```
Comment (ljgray) [inline draco/util/_fast_tools.pyx:377]: Commented stuff should be removed

File: draco/util/_fast_tools.pyx [hunk 2]
```diff
@@ -260,3 +271,207 @@ ctypedef fused real_or_complex:
...
+            means[ii] += arr[ii, jj]
+
+    for ii in range(nd):
+        means[ii] /= nsamp
+
+    for ii in range(nd):
+        for jj in range(nd):
+            out[ii, jj] = 0.0
+
+    #for iijj in prange(nd * nd, nogil=True):
```
Comment (ljgray) [inline draco/util/_fast_tools.pyx:419]: Same here

File: draco/util/_fast_tools.pyx [hunk 3]
```diff
@@ -260,3 +271,207 @@ ctypedef fused real_or_complex:
...
+    """
+
+    cdef int nd = arr.shape[0]
+    cdef int nsamp = arr.shape[1]
+
+    cdef int ii, jj
+
+    cdef real_or_complex* means = <real_or_complex*>malloc(sizeof(real_or_complex) * nd)
+
+    #for ii in prange(nd, nogil=True):
```
Comment (ljgray) [inline draco/util/_fast_tools.pyx:464]: Same here

File: draco/core/misc.py [hunk 4]
```diff
@@ -299,3 +299,11 @@ def setup(self, input_):
     def next(self, input_):
         """Immediately forward any input."""
         return input_
+
+
+class PassOn(task.MPILoggedTask):
```
Comment (ljgray) [inline draco/core/misc.py:304]: What's the use-case for this task?

File: draco/analysis/delay.py [hunk 5]
```diff
@@ -555,63 +556,91 @@ class DelaySpectrumEstimatorBase(task.SingleTask, random.RandomTask):
...
-            or self.average_axis not in ss.datasets[self.dataset].attrs["axis"]
-        ):
-            raise ValueError(
-                f"Specified axis to average over ({self.average_axis}) not in "
-                f"container of type {type(ss)}."
+        if not isinstance(datasets, list):
+            datasets = [datasets]
+            cross = False
+        else:
+            # cross = len(datasets) > 1
```
Comment (ljgray) [inline draco/analysis/delay.py:600]: curious why this check is disabled - wouldn't we want to use the cross spectrum only if there's more than one dataset?

**Comment by jrs65** [comment]
@ljgray this probably isn't worth spending too much time reviewing as it needs to be changed fairly dramatically to make it work with the other changes to the delay spectrum code. I probably won't have time to do that for quite a while.

**Comment by jrs65** [comment]
Superceded by #255

**Pull request closed without merging**